### DISCUSSION TITLE
Native Home app notification when the dishwasher has finished the program

### DISCUSSION THREAD

**Discussion by dlpgithub** [thoukydides/homebridge-homeconnect#229] [Q&A]
Is there a way to receive a native home app notification when the dishwasher has finished the program?
Something like this from the HomeConnect App:

![IMG_3915](https://github.com/thoukydides/homebridge-homeconnect/assets/48456028/065a169a-6e9a-46c6-b496-1410afe60998)

The dishwasher issues the following message in the HomeBridge event log:
[14/01/2024, 15:37:26] [HomeConnect] [Geschirrspüler] Event Program Finished

<img width="851" alt="Bildschirmfoto 2024-01-14 um 16 38 35" src="https://github.com/thoukydides/homebridge-homeconnect/assets/48456028/35c87425-1cf5-4769-8d49-22d41ed5409a">

**Comment by thoukydides** [comment]
The Apple Home app generates its own notifications for security-related events, e.g. doors being opened, doorbell pressed, or motion detected. It is not possible for a HomeKit accessory to generate additional events, or otherwise influence the Apple Home app's events.

Hence, the only way that this could be achieved would be by implementing a HomeKit service, such as a Door, and use that to trigger the notifications. Of course, the notification would be that the Door has been opened, rather than anything more relevant. This has been discussed previously, e..g in #38.

As you point out, the Home Connect app already generates these notifications, so it would be silly to have an Apple Home app notification with a completely irrelevant message. Hence, I will not be implementing this.

**Reply by dlpgithub** [reply]
I accept your answer, but i don't think it's silly.
I would prefer to receive a notification without the HomeConnect app 
and I don't want to install the manufacturer's app for every device.

For example, my robot vacuum cleaner sends a notification via the Home app when cleaning is complete.
So I thought that would be great for the dishwasher too.

**Comment by marksanctuary** [comment]
This feature request has also been on my mind for the Apple Home app support. I receive every Dishwasher Door Open/Closed notification which lets me know that my kids are working on loading the Dishwasher. But since the Appliance Active Program is built as a switch I can not add Home App Notifications to this switch.

To make this possible the Appliance Active Program would need to be reconfigured at a HomeKit Contact Sensor. Then the Apple Home app would allow notifications to be configured to let one know the Active Program was complete. My 2 cents, is that Appliance Active Program is a status and not really a switch. The Contact Sensor is closer to a status of the active program.

@thoukydides, thoughts?

**Comment by thoukydides** [comment]
Have you looked at the functionality that this plugin implements on the Active Program `Switch` service?

Copying from the [Wiki](https://github.com/thoukydides/homebridge-homeconnect/wiki/HomeKit-Mapping#switch-service-for-active-program

> ## `Switch` Service for Active Program
> 
> For most appliances that support programs a `Switch` service indicates when a program is currently running (including delayed start and in the process of aborting). It also allows the currently selected program to be started, or the active program to be stopped. For appliances that support pause and resume this also controls that functionality.
> 
> Turning the switch on while the active program is paused attempts to resume it if supported by the appliance. Otherwise, it attempts to start the selected program. For this to work most appliances require:
> * Connection to the Home Connect servers
> * Powered on
> * No active program
> * [Remote control](https://developer.home-connect.com/docs/api/status/remotecontrolactivationstate) enabled
> * [Remote program start](https://developer.home-connect.com/docs/api/status/remotestartallowancestate) enabled
> * No recent local (manual) control
> 
> Turning a switch off attempts to pause the active program if supported by the appliance and the appliance is not in an error state. Otherwise it stops the active program.
> 
> ### Extra Active Program `Switch` Characteristics
> 
> The `Status Active` and `Status Fault` characteristics are used to indicate the operation state of the appliance:
> * `Status Active` is cleared if a program is active but not running normally (e.g. when it is paused, requires user interaction, has an error, or is in the process of aborting). (Elgato Eve only shows this status when it is active.)
> * `Status Error` indicates when the appliance actually reports an error. (Elgato Eve only shows this status when it is active.)
> 
> For appliances that provide a countdown of the remaining program time this is reported in the `Remaining Duration` characteristic.
> 
> These characteristics are not part of the standard `Switch` service as defined by Apple, and are not supported by Apple's Home app. See the recommended [HomeKit Apps](https://github.com/thoukydides/homebridge-homeconnect/wiki/HomeKit-Apps) that should be used instead.

That's a lot more than just a sensor status. I also still cannot see the attraction of a generic notification such as "Contact Sensor Open" or "Door Closed" for a completely unrelated activity. I have tried hard to map Home Connect functionality to HomeKit with as close to appropriate semantics as possible, and intend to maintain that approach. So, no, I will not be changing the Active Program implementation.

(Hopefully Apple will eventually extend the HomeKit object model to support Matter 1.2, which should allow more functionality to be mapped with appropriate semantics...)

**Reply by marksanctuary** [reply]
Thank you for your thoughts. Yeah, unfortunately the switch seems to not provide the notifications option. And I can see that there is more function to that switch. 

Would you consider adding the Contact Sensor as addendum to the Active Program Switch?

I think this request would be very common to most users. It's a head scratcher why in the Home app users are unable to see when the dishwasher is done. Also its a head scratcher why Apple which is part of the Matter consortium, has not be adding the newer standards to HAP sooner, but that's another discussion all together.

As a power user I would use 3 party apps all day long, but I stick with the built in Apple apps for most things, especially with the home automation, since I have a whole family using the Home app. It really helps with my wife who is not a techie... aka. wife acceptance factor. LOL

I have attached screenshot capturing Apple's Home app shortcomings when using a switch vs a door/contact sensor.

![Status vs Status and Notifications](https://github.com/thoukydides/homebridge-homeconnect/assets/16604048/eec94cb3-2ac9-4c72-a9a0-3fa188f5f316)

**Reply by marksanctuary** [reply]
It would say something like, 
"The Dishwasher Active Program is Open",
"The Dishwasher Active Program is Closed",
... sigh ... Come on Apple!

You're consideration is helpful for this idea, but I do see why its not the best practice.

**Reply by marksanctuary** [reply]
The Matter 1.2 specification was released in October 2023. Web searched and found CES had many mentions of the new standards and products that work it them, including dishwashers. Scanned the OS updates (https://developer.apple.com/news/releases/) for the various Apple devices, no mention of HomeKit updates yet. 

I would imagine Apple is going to package it a part of a major developer homekit release during WWDC to be a part of the next major OS releases. I bet it will be 6 months to a year before the new standards waterfall down to Homebridge.

I am supportive either way. Hope this discussion also helps others looking for the same features too. Thanks for the hard work on the HomeConnect  plugin.

**Reply by marksanctuary** [reply]
Here is a funny one that hit me... I could rename the Dishwasher Activity with some "steam punk" feel in mind.

The Dishwasher Cycle Valve is Open
The Dishwasher Cycle Valve is Closed

As you can tell I would love to figure out a workaround until Apple does their part catching up HAP with the standards.

I was reviewing other discussions, and it seems you would greatly appreciate having all the standards available in both HAP and the HAP-Node.js interface. That way you would not have to defend the shortcomings of the current state things are in. The more I dig around about Matter and HAP the more I realize it's a bunch of tough issues.

**Comment by connor-g** [comment]
Not a direct answer, but you may be able to achieve something similar to your goal using the `homebridge-delay-switch` plugin.

- Create a switch called something like "Dishwasher Finished". Set the timer to 1ms. Then set the Trigger Sensor to Motion Sensor.
- In the Apple Home app, create an automation that when the Dishwasher program turns off, activate the Dishwasher Finished switch.
- Make sure you have notifications turned on for the Dishwasher Finished motion sensor.

With this setup, you should get a notification from the Home app with text like "Motion detected by Dishwasher Finished". It's not perfect, but pretty close.
